ecc/fourq/point.go

```diff
@@ -272,36 +272,48 @@ func (P *Point) Unmarshal(in *[Size]byte) bool {
 
 func (P *pointR1) IsOnCurve() bool {
 	t0, lhs, rhs := &Fq{}, &Fq{}, &Fq{}
+	// Check z != 0
+	eq0 := !P.Z.isZero()
 
+	// Check Eq 1: -X^2 + Y^2 == Z^2 + dT^2
 	fqAdd(t0, &P.Y, &P.X)    // t0  = y + x
 	fqSub(lhs, &P.Y, &P.X)   // lhs = y - x
 	fqMul(lhs, lhs, t0)      // lhs = y^2 - x^2
-	fqMul(rhs, &P.X, &P.Y)   // rhs = xy
-	fqSqr(rhs, rhs)          // rhs = x^2y^2
-	fqMul(rhs, rhs, &paramD) // rhs = dx^2y^2
-	t0.setOne()              // t0  = 1
-	fqAdd(rhs, rhs, t0)      // rhs = 1 + dx^2y^2
-	fqSub(t0, lhs, rhs)      // t0 = -x^2 + y^2 - (1 + dx^2y^2)
-	return t0.isZero()
```

Comment:
About commit `fourq: Handle the case of Z=0 for IsOnCurve and IsEqual.`

The original formula checked whether it matched the untwisted Edward curve, Ed : y² + x² = 1 + dx²y² (https://eprint.iacr.org/2015/625.pdf). I can guess what the new formulas are checking, what is it?

Reply:
The original formula does check the _affine_ twisted Edwards curve (`a=-1`), the changes now check the _projective_ version of the curve.